### ISSUE TITLE
[no-empty-interface] False positive when merging interface with class

### ISSUE THREAD

**Issue by IlyaSemenov** [typescript-eslint/typescript-eslint#1614]
**Repro**

```JS
module.exports = {
	root: true,
	extends: ['plugin:@typescript-eslint/recommended'],
}
```

```TS
interface Foo {
	name: string;
}

interface FooModel extends Partial<Foo> {}

class FooModel {
	boom(): void {
		console.log(this.name)
	}
}
```

**Expected Result**

The interface is allowed to exist. It is [merged](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) into the class.

**Actual Result**

The interface is replaced with:

```ts
type FooModel = Partial<Foo>
```

which leads to two errors:

* Duplicate identifier 'FooModel'. ts(2300)
* Property 'name' does not exist on type 'FooModel' .ts(2339)

**Additional Info**

Merging an interface into a class is a relatively widely used pattern, e.g. when using [Objection.js](https://vincit.github.io/objection.js/) ORM.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.19.2` |
| `@typescript-eslint/parser`        | `2.19.2` |
| `TypeScript`                       | `3.7.5` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `10.15.3` |
| `npm`                              | `6.9.0` |

**Comment by bradzacher** [comment]
> Merging an interface into a class is a relatively widely used pattern

I would call that an objectively false statement. Declaration merging itself is not even a widely used feature, let alone the weird case of merging a class and an interface.

Why so you need to merge an interface and a class together?
Why not use implements or declare properties?

**Comment by IlyaSemenov** [comment]
> I would call that an objectively false statement.

Same as with the second issue, it doesn't really matter. Declaration merging is a documented feature, I don't see any reason why the linter should break the working code.

> Why not use implements or declare properties?

Because it would be violating DRY. Should I really explain why duplicating the code is not good?

```ts
// Interface is to be used when the data is serialized and passed over the network
interface Foo {
  name: string;
  something: string;
  more: string;
  and_more: string;
  // Please don't forget to update the model when you edit fields here.
}

// Model instance is to be used when working with ORM
class FooModel {
  name: string;
  something: string;
  more: string;
  and_more: string;
  // Please don't forget to edit the interface when you edit fields here.

  boom(): void {
    console.log(this.name)
  }
}
```

How good is that code?

**Comment by bradzacher** [comment]
Same with the other issue, I often push back so that I can understand the use case.
Sometimes it's because it's a smelly thing, and in understanding the intention, I can educate the "correct" way to do it.
Similarly, sometimes there's nothing explicitly wrong with their code, but there is just a better way to do things (i.e. cases where people have used `any` instead of `never` for generics).

And sometimes (like in this case) it's because it's a pattern I'm unfamiliar with, so I want to understand what's being done and why, so I can figure out the best way to fix an issue.

I haven't ever _had_ to use declaration merging, so I like to understand why people want to use it.
I often discourage it because it's not an easy thing for everyone to understand.
In the case of merging a class and an interface, it has corner cases that can lead to type safety.
But that's neither here nor there in this case.

---

This has been unsupported (as declaration merging is mostly unsupported in all of our rules) because it's non-trivial to add in, and is not a very widely used TS feature.

It will probably be a bit hard to add in, because it requires scope analysis, but happy if you want to add this in behind a flag.

**Comment by IlyaSemenov** [comment]
One more use case is treating Nuxt.js (a Vue server-side rendering framework) `asyncData`:

```vue
<script>
interface AsyncData {
	foo: string
	bar: number
}

interface MyPage extends AsyncData {
	// merge
}

@Component({
	async asyncData({ app, error }): Promise<AsyncData> {
		const { foo } = await app.$axios.$get<FooApiResponse>(`/api/foo`)
		if (!foo) {
			return error({ message: 'invalid foo' })
		}
		return { foo, bar: 123 } // type check works here
	},
})
class MyPage extends Vue {
	hello = 'hello'

	boom() {
		console.log(this.hello + this.foo)
	}
}

export default MyPage
</script>

<template>
  <div>bar = {{ bar }}</div>
</template>
```

On the one hand, everything returned by `asyncData` is merged into `this` once a component is instantiated (this is how Nuxt.js works). On the other hand, one would also want to explicitly define `asyncData` return type to enable type check on the returned object.

Granted, the syntax is far from elegant, but I'd call it arguably lesser evil than copy/paste.

**Comment by bradzacher** [comment]
The problem with interface + class merging is that it is pretty unsafe.

```ts
interface AsyncData {
	foo: string
	bar: number
}
interface MyPage extends AsyncData {
	// merge
    someOtherPropertyThatIsntInAsyncData: string;
}
@Component({ ... })
class MyPage extends Vue {
	hello = 'hello'

	boom() {
        // compile-time safe due to declaration merging
		return this.someOtherPropertyThatIsntInAsyncData.length;
	}
}

const x = new MyPage();
x.boom(); // runtime error: Cannot read property 'length' of undefined
```
This could also happen if your `AsyncData` defines a property which is not returned by your API.

The example you gave in your OP is safe purely because your merged interface `extends Partial<Foo>`.
Which is why I hesitate to recommend interface+class decl merging; it's an easy footgun.

---

But it is valid code, so I'm happy to accept a PR to support it.

**Comment by Samuel-Therrien-Beslogic** [comment]
A use case for us: self-referencing type aliases:
```ts
// Type alias 'TranslationKeys' circularly references itself. ts(2456)
// 'any' overrides all other types in this union type. eslint(@typescript-eslint/no-redundant-type-constituents)
type TranslationKeys = Record<string, TranslationKeys | string>
```
vs
```ts
// An interface declaring no members is equivalent to its supertype. eslint(@typescript-eslint/no-empty-interface)
interface TranslationKeys extends Record<string, TranslationKeys | string> {}
```

**Comment by bradzacher** [comment]
@Samuel-Therrien-Beslogic that is unrelated to this issue. This issue is about declaration merging an interface with a class.

**Comment by Josh-Cena** [comment]
In v8, we are merging `no-empty-interface` into a bigger rule called `no-empty-object-type` to make the intent clearer that we are only interested in banning `{}` as a type, but not preventing you from using the interface syntax to create aliases so you can then do merging. You should not be worried to turn on the `allowSingleExtends` option (or `allowInterfaces: "with-single-extends"` in the new rule).

Realistically speaking, this use case is impossible to detect reliably without using type information, so allowing `interface X extends Y {}` is the best we can provide anyway.
